src/main/java/com/blazemeter/taurus/junit/generator/Supervisor.java

```diff
@@ -0,0 +1,187 @@
+    private final Reporter reporter;
+    private final ThreadCounter counter;
+
+    private int concurrency;
+    private int steps;
+    private long iterations;
+    private float rampUp;
+    private float hold;
+
+    private volatile boolean isInterrupted = false;
+    private final AtomicBoolean isStopped = new AtomicBoolean(false);
+
+    public Supervisor(List<Class> classes, Properties properties) {
+        this.properties = properties;
+        this.classes = classes;
+        try {
+            this.reporter = new TaurusReporter(properties.getProperty(REPORT_FILE));
+        } catch (IOException e) {
+            log.log(Level.SEVERE, "Failed to create reporter", e);
+            throw new CustomRunnerException("Failed to create reporter", e);
+        }
+        this.counter = new Counter();
+        initParams(properties);
+        addShutdownHook();
+    }
+
+    private void addShutdownHook() {
+        Runtime.getRuntime().addShutdownHook(new Thread() {
+            @Override
+            public void run() {
+                isInterrupted = true;
+                stopWorkers();
+            }
+        });
+    }
+
+
+    private void initParams(Properties props) {
+        concurrency = Integer.valueOf(props.getProperty(CONCURRENCY, "1"));
+        steps = Integer.valueOf(props.getProperty(STEPS, String.valueOf(Integer.MAX_VALUE)));
+        rampUp = Float.valueOf(props.getProperty(RAMP_UP, "0"));
+        hold = Float.valueOf(props.getProperty(HOLD, "0"));
+        iterations = Long.valueOf(props.getProperty(ITERATIONS, "0"));
+        if (iterations == 0) {
+            if (hold > 0) {
+                iterations = Long.MAX_VALUE;
+            } else {
+                iterations = 1;
+            }
+        }
+    }
+
+    private void createWorkers() {
+        for (int i = 0; i < concurrency; i++) {
+            Worker worker = new Worker(classes, properties, reporter, counter, getWorkerDelay(i), iterations);
+            worker.setName("Worker #" + i);
+            worker.setDaemon(false);
+            workers.add(worker);
+        }
+    }
+
+    public void execute() {
+        createWorkers();
+        long workingTime = (long) (rampUp + hold) * 1000;
+        long endTime = (workingTime == 0) ? 0 : (System.currentTimeMillis() + workingTime);
+        for (Worker w : workers) {
+            w.start();
+        }
+        waitForFinish(endTime);
+    }
+
+    private void waitForFinish(long endTime) {
+        while (true) {
+            if (isInterrupted) {
+                log.fine("Supervisor was interrupted. Break loop.");
+                stopWorkers();
+                break;
+            }
+
+            long currTime = System.currentTimeMillis();
+            if (0 < endTime && endTime <= currTime) {
+                log.info("Duration limit reached, stopping");
+                stopWorkers();
+                break;
+            }
+
+            if (0 == counter.getActiveThreads()) {
```

Comment:
You can't rely on concurrency counter here. You have to rely on real alive threads, because theads may go inactive because of  exception.